### DISCUSSION TITLE
Pyvista Polydata to Trimesh Mesh Object Conversion

### DISCUSSION THREAD

**Discussion by venugovh** [pyvista/pyvista#2268] [Q&A] [answered]
Hi,

Is there a way to convert a `pyvista `mesh object to a `trimesh `mesh object? I think `pyvista.wrap` does the vice-versa.

Thanks!

**Comment by whophil** [comment]
There is a two-liner in one of the ray-casting filters which does this.

https://github.com/pyvista/pyvista/blob/28914fc2236cd5850f536b151d0ef383855d8519/pyvista/core/filters/poly_data.py#L1948-L1949

It requires that your mesh be a `PolyData` which is all triangles, which can be accomplished with:

```python
mesh = mesh.extract_surface().triangulate()
```

**Comment by olly-writes-code** [comment]
Adding a super simple example as the syntax has changed slightly for newer versions of pyvista

``` python
from pyvista import examples
import trimesh

dataset = examples.download_cow()
cow = dataset.rotate_x(90) ### it's only fair that the cow gets to stand upright

cow_tri = cow.triangulate()
faces_as_array = cow_tri.faces.reshape((cow_tri.n_faces_strict, 4))[:, 1:] 
tmesh = trimesh.Trimesh(cow_tri.points, faces_as_array) 
tmesh.show()
```

**Reply by user27182** [reply]
Thanks for the tips! You can also do `faces_as_array = cow_tri.regular_faces` to avoid the need to reshape.

It would be nice to encapsulate this into a new `to_trimesh` method.

**Reply by user27182** [reply]
> It would be nice to encapsulate this into a new `to_trimesh` method. 

Dedicated to/from trimesh functions are added in #8176 in case anyone wants to test out or review the proposed API.

**Comment by user27182** [comment, marked as answer]
There is a new API for converting to/from Trimesh:
https://dev.pyvista.org/api/utilities/_autosummary/pyvista.to_trimesh
https://dev.pyvista.org/api/utilities/_autosummary/pyvista.from_trimesh

It will be part of the next PyVista release (v0.47).

It is recommended to use these for converting Trimesh meshes.

I am marking this as the new accepted answer for users to find in the future.

**Reply by banesullivan** [reply]
For posterity, also published a new trimesh accessor to streamline using pyvista and trimesh here: https://github.com/pyvista/pyvista-trimesh
